typescript/packages/jumble/integration/smoke-test.ts

```diff
@@ -1,23 +1,69 @@
 import { launch } from "@astral/astral";
 import { assert } from "@std/assert";
-import { login } from "./utils.ts";
+import { addCharm, inspectCharm, login } from "./utils.ts";
 
 const FRONTEND_URL = "http://localhost:5173/";
+const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
 
 async function main() {
-  const browser = await launch();
-  console.log(`Waiting to open website at ${FRONTEND_URL}`);
-  const page = await browser.newPage(FRONTEND_URL);
-  console.log(`Opened website at ${FRONTEND_URL}`);
-
-  await login(page);
-
-  const anchor = await page.waitForSelector("nav a");
-  assert(
-    (await anchor.innerText()) === "common-knowledge",
-    "Logged in and Common Knowledge title renders",
-  );
-  await browser.close();
+  const { charmId, space } = await addCharm();
+  console.log(`Charm added`, { charmId, space });
+
+  let browser;
+  try {
+    browser = await launch();
+    console.log(`Waiting to open website at ${FRONTEND_URL}`);
+    const page = await browser.newPage(FRONTEND_URL);
+    console.log(`Opened website at ${FRONTEND_URL}`);
+
+    console.log("Logging in");
+    await login(page);
+
+    console.log("Checking if logged in");
+    const anchor = await page.waitForSelector("nav a");
+    assert(
+      (await anchor.innerText()) === "common-knowledge",
+      "Logged in and Common Knowledge title renders",
+    );
+
+    await page.goto(`${FRONTEND_URL}${space}/${charmId}`);
+    console.log(`Waiting for charm to render`);
+
+    await sleep(5000);
+    await page.waitForSelector("a[aria-current='charm-title']");
+    const el = await page.$("a[aria-current='charm-title']");
+    console.log(await el?.innerText());
+
+    assert(
+      (await el!.innerText()) === "Simple Value: 1",
+      "Charm renders",
+    );
+
+    console.log("Clicking button");
+    const button = await page.waitForSelector(
+      "div[aria-label='charm-content'] button",
+    );
+    await button.click();
+
+    // FIXME(ja): sleep to let changes propagate to remote storage
+    await sleep(5000);
```

Comment:
If we need to wait here, lets add a utility that polls for some state every ~100ms and times out after awhile. Otherwise, if 5000ms is good some times, then it's probably not sufficient other times, introducing a race

If `waitForSelector()` is insufficient e.g. we're waiting for an element with a specific value, maybe the util can do something like:

```ts
export const waitForElementWithText = (page: Page, selector: string, text: string, options) => {
  for (let i = 0; i < (options?.retries || 10); i++) {
    const el = await page.$(selector);
    if (!el) {
      await sleep(options?.delay);
      continue;
    }
    if ((await.el.innerText()) === text) {
      return;
    }
  }
  throw new Error("");
}
```

Reply:
Agreed. The challenge is that I don't think we have a element that we can trust shows us whether the data propagated out or not yet. the green/yellow circle only updates every five seconds. Perhaps we can change that.